golang/x/relay/keeper/requests.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/summa-tx/relays/golang/x/relay/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (k Keeper) emitProofRequest(ctx sdk.Context, pays, spends types.Hash256Digest, paysValue, id uint64) {
+	ctx.EventManager().EmitEvent(types.NewProofRequestEvent(pays, spends, paysValue, id))
+}
+
+func (k Keeper) getRequestStore(ctx sdk.Context) sdk.KVStore {
+	return k.getPrefixStore(ctx, types.RequestStorePrefix)
+}
+
+func (k Keeper) getRequestIDStore(ctx sdk.Context) sdk.KVStore {
+	return k.getPrefixStore(ctx, types.RequestIDStorePrefix)
+}
+
+func (k Keeper) hasRequest(ctx sdk.Context, id []byte) bool {
+	store := k.getRequestStore(ctx)
+	return store.Has(id)
+}
+
+func (k Keeper) setRequest(ctx sdk.Context, spends []byte, pays []byte, paysValue uint64, numConfs uint8) sdk.Error {
+	store := k.getRequestStore(ctx)
+
+	valid := k.validateRequests(spends, pays)
+	if !valid {
+		return types.ErrInvalidRequest(types.DefaultCodespace)
+	}
+
+	spendsDigest := btcspv.Hash256(spends)
+	paysDigest := btcspv.Hash256(pays)
+
+	request := types.ProofRequest{
+		Spends:      spendsDigest,
+		Pays:        paysDigest,
+		PaysValue:   paysValue,
+		ActiveState: true,
+		NumConfs:    numConfs,
+	}
+
+	// When a new request comes in, get the id and use it to store request
+	id := k.getID(ctx)
+
+	buf, err := json.Marshal(request)
+	if err != nil {
+		return types.ErrMarshalJSON(types.DefaultCodespace)
+	}
+	store.Set(id, buf)
+
+	// Increment the ID
+	k.incrementID(ctx)
+
+	// Emit Proof Request event
+	numID := binary.BigEndian.Uint64(id)
+	k.emitProofRequest(ctx, request.Pays, request.Spends, request.PaysValue, numID)
+	return nil
+}
+
+func (k Keeper) getRequest(ctx sdk.Context, id []byte) types.ProofRequest {
+	store := k.getRequestStore(ctx)
+	buf := store.Get(id)
+	var request []types.ProofRequest
+	json.Unmarshal(buf, &request)
+	return request[0]
+}
+
+func (k Keeper) incrementID(ctx sdk.Context) {
+	// TODO: Is there a better way of incrementing this? Have to store as bytes...
+	store := k.getRequestIDStore(ctx)
+	// get id
+	id := k.getID(ctx)
+	// convert id to uint64 and add 1
+	newID := binary.BigEndian.Uint64(id) + 1
+	// convert back to bytes and store
+	b := make([]byte, 8)
+	binary.BigEndian.PutUint64(b, newID)
+	store.Set([]byte("id"), b)
+}
+
+func (k Keeper) getID(ctx sdk.Context) []byte {
+	store := k.getRequestIDStore(ctx)
+	id := []byte("id")
+	if !store.Has(id) {
+		store.Set(id, []byte{0})
+	}
+	return store.Get(id)
+}
+
+func (k Keeper) validateRequests(spends []byte, pays []byte) bool {
+	if len(spends) != 36 {
+		return false
+	}
+	if len(pays) > 50 {
+		return false
+	}
+	return true
+}
+
+func (k Keeper) checkRequests(ctx sdk.Context, reqIndices uint16, vin []byte, vout []byte, requestID []byte) bool {
+	// TODO: Add errors
+	if !btcspv.ValidateVin(vin) {
+		return false
+	}
+	if !btcspv.ValidateVout(vout) {
+		return false
+	}
+
+	inputIndex := uint8(reqIndices >> 8)
```

Comment:
let's go ahead and split these up into 2 uint8s instead of 1 uint16. they're packed because solidity sucks

Reply:
Is this what you mean? ```func (k Keeper) checkRequests(ctx sdk.Context, inputIndex, outputIndex uint8, vin []byte, vout []byte, requestID []byte) bool {```

Reply:
Yes

Reply:
right now you get something like `0xff00` which gets unpacked into `0xff` and `0x00`. We did this because solidity stack operations are limited to depth 16, so having 1 less argument was important.

Go is not limited that way